### ISSUE TITLE
cargo check --test foo compiles too much

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/cargo#4059]
From a local compilation:

```
$ cargo new foo
$ cd foo
$ mkdir tests
$ touch tests/a.rs
$ cargo check --test a -v
   Compiling foo v0.1.0 (file:///home/alex/code/foo)
     Running `rustc --crate-name foo src/lib.rs --crate-type lib --emit=dep-info,link -C debuginfo=2 -C metadata=6d607d3e3579f789 -C extra-filename=-6d607d3e3579f789 --out-dir /home/alex/code/foo/target/debug/deps -L dependency=/home/alex/code/foo/target/debug/deps`
     Running `rustc --crate-name a tests/a.rs --emit=dep-info,link -C debuginfo=2 --test -C metadata=e8ff428dccc81b86 -C extra-filename=-e8ff428dccc81b86 --out-dir /home/alex/code/foo/target/debug/deps -L dependency=/home/alex/code/foo/target/debug/deps --extern foo=/home/alex/code/foo/target/debug/deps/libfoo-6d607d3e3579f789.rlib`
    Finished dev [unoptimized + debuginfo] target(s) in 0.32 secs
```

Notably nothing was compiled with metadata as an output.

cc @nrc

**Comment by BurningMind** [comment]
After investigating this issue, it seems that the problem comes from [this line in `cargo_compile.rs`](https://github.com/rust-lang/cargo/blob/master/src/cargo/ops/cargo_compile.rs#L597):

```rust
targets.append(&mut propose_indicated_targets(pkg, tests, "test", Target::is_test, test)?);
```

This line adds integration tests that match the `--test X` (or `--tests`) flag to the list of targets to be built. However, the profile passed to that function (the last argument) is the `test` profile, which has `check == false`. Hence code gets generated.

This could be fixed by passing the `check_test` profile from PR #4039 when we are checking, and `test` otherwise.

**Comment by BurningMind** [comment]
Should I take care if this issue in the same PR or should I open another one once the first is closed?

**Comment by alexcrichton** [comment]
@BurningMind oh feel free to take care of this issue in your PR! No need for a separate one.

**Comment by ehuss** [comment]
This is fixed by #4592.
